provider/server/exec.go

```diff
@@ -124,21 +128,35 @@ func (e executorExec) Exec(ctx context.Context, in io.Reader, out io.WriteCloser
 		return fmt.Errorf("machine %s has not yet been synced", machineID)
 	}
 
-	uri, err := e.Libvirt.ConnectGetUri()
+	domainXMLData, err := e.Libvirt.DomainGetXMLDesc(domain, 0)
 	if err != nil {
-		return fmt.Errorf("error getting connection uri")
+		return fmt.Errorf("failed to lookup domain: %w", err)
+	}
+
+	domainXML := &libvirtxml.Domain{}
+	if err := domainXML.Unmarshal(domainXMLData); err != nil {
+		return fmt.Errorf("failed to unmarshal domain: %w", err)
 	}
 
-	cmd := exec.CommandContext(ctx, e.VirshExecutable, "-c", strings.TrimSpace(uri), "console", machineID)
+	if domainXML.Devices == nil || len(domainXML.Devices.Consoles) == 0 {
+		return errors.New("device console not set in machine domainXML")
+	}
+	ttyPath := domainXML.Devices.Consoles[0].TTY
 
-	f, err := pty.Start(cmd)
```

Comment:
why is permission 0? i could find any information about this setup. How is it working?

Reply:
When opening an existing file for both reading and writing (O_RDWR) without the intention of creating a new file (O_CREATE not set), the permission parameter is disregarded. The file's current permissions are utilized, rendering the provided permission value ineffective.

Therefore, it is set to 0 in this case.